matrix-mult.futil

```diff
@@ -0,0 +1,234 @@
+component main() -> () {
+  cells {
+    @external(1) mem1 = comb_mem_d2(32, 2, 2, 1, 1); //matrix 1, and will store the result
+    @external(1) mem2 = comb_mem_d2(32, 2, 2, 1, 1); //matrix 2
+    val = std_reg(32); // holds one number of the result matrix
+    val1 = std_reg(32);
+    val2 = std_reg(32);
+    val3 = std_reg(32);
+    val4 = std_reg(32);
+    val5 = std_reg(32);
+    val6 = std_reg(32);
+    val7 = std_reg(32);
+    val8 = std_reg(32);
+    temp1 = std_reg(32);
+    temp2 = std_reg(32);
+
+    add = std_add(32);
+    mul = std_mult_pipe(32);
+  }
+
+  wires {
+    group read1 { // read value 00 from matrix 1
+      mem1.addr0 = 1'b0;
+      mem1.addr1 = 1'b0;
+      val1.in = mem1.read_data;
+      val1.write_en = 1'b1;
+      read1[done] = val1.done;
+    }
+
+    group read2 { // read value 00 from matrix 2
+      mem2.addr0 = 1'b0;
+      mem2.addr1 = 1'b0;
+      val2.in = mem2.read_data;
+      val2.write_en = 1'b1;
+      read2[done] = val2.done;
+    }
+
+
+    group read3 { //read value 01 from matrix 1
+      mem1.addr0 = 1'b0;
+      mem1.addr1 = 1'b1;
+      val3.in = mem1.read_data;
+      val3.write_en = 1'b1;
+      read3[done] = val3.done;
+    }
+
+    group read4 { // read value 01 from matrix 2
+      mem2.addr0 = 1'b0;
+      mem2.addr1 = 1'b1;
+      val4.in = mem2.read_data;
+      val4.write_en = 1'b1;
+      read4[done] = val4.done;
+    }
+
+    group read5 { // read value 10 from matrix 1
+      mem1.addr0 = 1'b1;
+      mem1.addr1 = 1'b0;
+      val5.in = mem1.read_data;
+      val5.write_en = 1'b1;
+      read5[done] = val5.done;
+    }
+
+    group read6 { // read value 10 from matrix 2
+      mem2.addr0 = 1'b1;
+      mem2.addr1 = 1'b0;
+      val6.in = mem2.read_data;
+      val6.write_en = 1'b1;
+      read6[done] = val6.done;
+    }
+
+    group read7 { // read value 11 from matrix 1
+      mem1.addr0 = 1'b1;
+      mem1.addr1 = 1'b1;
+      val7.in = mem1.read_data;
+      val7.write_en = 1'b1;
+      read7[done] = val7.done;
+    }
+    group read8 { // read value 11 from matrix 2
+      mem2.addr0 = 1'b1;
+      mem2.addr1 = 1'b1;
+      val8.in = mem2.read_data;
+      val8.write_en = 1'b1;
+      read8[done] = val8.done;
+    }
+
+    group mul_upd { //multiply 00 x 00, store in temp1
+      mul.left = val1.out;
+      mul.right = val2.out;
+      mul.go = !mul.done ? 1'd1;
+      temp1.in = mul.out;
+      temp1.write_en = mul.done;
+      mul_upd[done] = temp1.done;
+    }
+
+    group mul_upd2 { //multiply 01 x 10, store in temp2
+      mul.left = val3.out;
+      mul.right = val6.out;
+      mul.go = !mul.done ? 1'd1;
+      temp2.in = mul.out;
+      temp2.write_en = mul.done;
+      mul_upd2[done] = temp2.done;
+    }
+
+    group upd { // add together and store in val
+      add.left = temp1.out;
+      add.right = temp2.out;
+      val.in = add.out;
+      val.write_en = 1'b1;
+      upd[done] = val.done;
+    }
```

Comment:
Ah very clever, so the same `upd` can be used repeatedly as long as `temp1` and `temp2` have been set up correctly!

Reply:
This way is really great, but when we meet in person later I'd love to also show you another possibility: one group that uses two computational cells. Not necessarily what you'll want here, but I just want to show you the technique!